### DISCUSSION TITLE
transfer   如何调用  addToRight  ， addToLeft 方法

### DISCUSSION THREAD

**Discussion by 8502596** [element-plus/element-plus#18152] [General]
##   element-ui   中 transfer 可以用如下代码实现 不点击button 实现数据穿梭  请问如何在 element-plus 中实现这一点

``` vue


<template>
  <el-transfer
      ref="reserve"
      filterable
      :filter-method="filterMethod"
      filter-placeholder="请输入城市拼音"
      @left-check-change="handleChange"
      v-model="value"
      :data="data">
  </el-transfer>
</template>

<script>
export default {
  data() {

    const generateData = () => {
      const data = [];
      const cities = ['上海', '北京', '广州', '深圳', '南京', '西安', '成都'];
      const pinyin = ['shanghai', 'beijing', 'guangzhou', 'shenzhen', 'nanjing', 'xian', 'chengdu'];
      cities.forEach((city, index) => {
        data.push({
          label: city,
          key: index,
          pinyin: pinyin[index]
        });
      });
      return data;
    };
    return {
      data: generateData(),
      value: [],
      filterMethod(query, item) {
        return item.pinyin.indexOf(query) > -1;
      }
    };
  },

  methods: {
    handleChange(value, direction, movedKeys) {
      // console.log(value, direction, movedKeys);
      this.$nextTick(() => {
        console.log(this.$refs.reserve)//这个就是它包含的所有的属性以及事件，如果需要别的操作直接执行，也可以查询到
        this.$refs.reserve.addToRight();//直接执行到右事件
      });

    }
  }
};


``` 

![image](https://github.com/user-attachments/assets/1bf85f36-abce-4b41-a6b1-d5e3b3ac293f)

**Comment by warmthsea** [comment]
https://github.com/element-plus/element-plus/pull/18162

**Reply by 8502596** [reply]
是否 能添加   获取  左侧 选中的数据  及 右侧选中的数据 

@warmthsea 

![image](https://github.com/user-attachments/assets/bfbc5f30-07ff-4e4b-8b8f-f290a85b42e2)

**Reply by 8502596** [reply]
以及能否添加 checked 之后 后面的处理阻断，例如选中了某项  需要确认  穿梭按钮 才能生效，或者干脆让穿梭按钮不生效，例如 如下 代码  选择 option 1 会不生效

```  vue

<template>
  <el-transfer ref="transferRef" v-model="value" :data="data" @left-check-change="left" />
  <div class="mt-10">
      <el-button @click="testLeft()">testLeft</el-button>
      <el-button @click="testRight()">testRight</el-button>
  </div>

  <div class="mt-10">
      <el-button @click="getLeftData()">getLeftData</el-button>
      <el-button @click="getRightData()">getRightData</el-button>
  </div>
</template>

<script lang="ts" setup>
import { ref,nextTick } from 'vue'
import {TransferKey} from "element-plus";
const left =async (value: TransferKey[], movedKeys?: TransferKey[]) =>{
  if(movedKeys[0]==1)
  {
    let buttons = document.querySelectorAll('.el-transfer__button');
    await nextTick();
    let button=buttons[buttons.length-1] as HTMLButtonElement;
    button.disabled=true;   //禁用
    button.classList.add("is-disabled"); // 添加禁用样式
  }
}

const transferRef = ref()
function testRight() {
  transferRef.value.addToRight()
}

function testLeft() {
  transferRef.value.addToLeft()
}

function getLeftData() {
  console.log(transferRef.value.sourceData)
}

function getRightData() {
   console.log(transferRef.value.targetData)
}


interface Option {
  key: number
  label: string
  disabled: boolean
}

const generateData = () => {
  const data: Option[] = []
  for (let i = 1; i <= 15; i++) {
    data.push({
      key: i,
      label: `Option ${i}`,
      disabled: i % 4 === 0,
    })
  }
  return data
}

const data = ref<Option[]>(generateData())
const value = ref([])
</script>


``` 
我认为  left-check-change 函数应该设计为 可以阻断组件继续执行  是否有个返回值 如果true  组件才会继续进行 否则则阻断，在代码中  因为我是新手不知道  有没有更好的方式，以及阻断之后需要异步等待 例如  选中之后 弹窗窗口确认之后 才能点穿梭按钮或者是不被选中。描述中如果存在问题 请见谅

![image](https://github.com/user-attachments/assets/cc0ad04b-4561-4164-be66-8455df5af3a8)

![image](https://github.com/user-attachments/assets/1a42374d-86e9-4392-9b8e-9450c9559a78)

**Reply by warmthsea** [reply]
> ```
> left 
> ```

这个 pr 一并抛出了

**Reply by warmthsea** [reply]
@8502596 

关于中间确认的操作我觉得是很好的点子，我会考虑实现它

**Reply by 8502596** [reply]
您好，我的意思是 左侧和右侧  checked  选中的数据

而新添加的 sourceData 和  targetData 则是 整个面板的数据   里面包含了  被选中的数据  获取这些被选中的数据似乎没有接口

只能通过 left-checked-change  事件中获取，但如果不在这个事件中 则不太好获取 例如 前端设置左侧选中数据 及 右侧 选中数据

传递到后端。
![image](https://github.com/user-attachments/assets/3f80b5e5-93cc-45e0-a042-2f9fd9ed99ba)

如图  选中的数据个数  应该是 
leftcheckedData ,  rightcheckedData
4  ， 0

**Reply by 8502596** [reply]
我有一个想法   checkbox  选中时  调用  Popconfirm 气泡确认框  以及 Popconfirm 自定义内容使用  Tabs 标签页，这看着有些复杂，但实现的页面效果却非常简单，请问该如何实现呢，主要是 el-transfer 的各项checkbox 如何 添加 Popconfirm 或者是Popover 气泡卡片 带 确认或取消按钮  ，顺便说一下 Popconfirm   应该以 checkbox 或者 radiobox 为示例吧，确认了 就是选中了。

![image](https://github.com/user-attachments/assets/2f1b9d56-c862-4267-ace2-d04f487cd8cc)


![1725846033505](https://github.com/user-attachments/assets/1a7d0d2a-ae27-4b36-8000-dddbc751fb93)

![1725846098488](https://github.com/user-attachments/assets/ab75f70d-9a60-4eb0-b87d-66f7bdea245b)

![1725846164062](https://github.com/user-attachments/assets/718ddcba-f3b4-4b41-8e43-7545ee82cac6)


进一步  pannel 里面的组件  应该 可以是 checkbox ， radiobox ，input  ，label  等等 

![image](https://github.com/user-attachments/assets/724ba4c8-76da-4336-bf1f-dbd35bc0b702)

他应该是 任何 components  之中能显示的内容  例如 button 应该 显示为  button  而不是  他的label   ，这个组件有意思的点在于   组件 外层的 是 2个 pannel  2个 button  ，而pannel 内  应该是内置的 checkbox  我并没有看源码，我认为 将  内置的checkbox  换成 任意的 
components   是可行的  ，虽然大部分情况下  后端只是需要 一个  数组包含 label  或value 的值 ，但 页面展示  可以是radiobox  可以是 button  可以 是label   他们都可以有对应的  key  label  disable  状态等。

而 上面提到的  选中 弹出  popconfirm popover  只需要 给他传入 自定义组件内容即可  类似java 里面的接口  或抽象类，请原谅我对前端一无所知，只是个人的一点想法。



例如  前端完全可以 不需要这个checkbox  这个确认按钮   那么他只需要对应的 text 组件 即可
但 text 组件似乎没有被选中 这个属性  那么应该可以自己实现  即  自定义组件 实现初始化，点击 选中，和取消后的 各种属性
然后交由 transfer 组件进行显示  
![image](https://github.com/user-attachments/assets/c0dca72b-9614-408a-8a52-c5f6ba0761d6)

我不知道如何具体实现  但如果是一个select  其实情况是与  transfer 组件 类似的
![image](https://github.com/user-attachments/assets/6f79ceab-2584-4159-83ca-b7d63faf9c12)

我的意思是 transfer  应该更基础能提供各种 类型，否则 似乎与 select 功能相重复了  只需将select  下拉列表不关闭  弄2个 select 
即 实现了 transfer  大部分功能  我想 transfer  应该是一个更基础 更容易扩展的组件  例如 可以传入 radiobox   button 等这样 组件极大的扩展了功能，另外 我需要做一个  three  transfer   逻辑功能其实非常简单  即实现 A+B=C  将2个集合  可视化
类似与这样 

![image](https://github.com/user-attachments/assets/08a0715e-f249-43e8-bbee-01947814bcb6)

这其实是一个非常简单的 前端应用  但我目前用  许多框架无法简单的实现 ，我准备用 2个 transfer  将第一个的 右侧面板和 按钮隐藏  加上 addToRight  功能及  获取列表  这样 即可 简单 的实现 这一应用，请问 还有简单的实现方法吗？

下一步 checkbox 选中时弹窗 以确保选中的选项  或者 选中的选项  需要保存对应的数据  于是有了上面  transfer 需要扩展更多组件的想法，虽然过程复杂 但实现的应用  应该说是很常用 很正常的逻辑 

![image](https://github.com/user-attachments/assets/203b1a6f-09fb-4b37-8ce1-ab346fd31d15)

![image](https://github.com/user-attachments/assets/898553c2-0ed7-4b07-b633-161742f8981a)

类似于这样  选中 然后弹窗确认  我只需将  transfer  checkbox 换成  Popconfirm, Popover  组件即可完成最终的应用 弹出对话框确认及保存数据   当然前提是 el-transfer 得支持 前面 提到的支持其他组件
